calculator/mod_tools.go

```diff
@@ -25,3 +31,120 @@ func MOD(modName string, modType mod.Type, modVal any) mod.Mod {
 	}
 	return realMod
 }
+
+func buildModListForNodeList(env *Environment, nodes map[string]data.Node) *ModList { // TODO ADD: finishJewels)
+	// Initialise radius jewels
+	/* *
+	for _, rad in pairs(env.radiusJewelList) do
+		wipeTable(rad.data)
+		rad.data.modSource = "Tree:"..rad.nodeId
+	end
+	/* */
+
+	// Add node modifiers
+	var modList = NewModList()
+	for _, node := range nodes {
+		var nodeModList = buildModListForNode(env, node)
+		modList.AddDB(nodeModList)
+
+		// TODO: Is this still a good idea?
+		/* *
+		if env.Mode == "MAIN" {
+			node.finalModList = nodeModList
+		}
+		/* */
+	}
+
+	/* *
+	// TODO
+	if finishJewels then
+		// Process extra radius nodes; these are unallocated nodes near conversion or threshold jewels that need to be processed
+		for _, node in pairs(env.extraRadiusNodeList) do
+			local nodeModList = calcs.buildModListForNode(env, node)
+			if env.mode == "MAIN" then
+				node.finalModList = nodeModList
+			end
+		end
+
+		-- Finalise radius jewels
+		for _, rad in pairs(env.radiusJewelList) do
+			rad.func(nil, modList, rad.data)
+			if env.mode == "MAIN" then
+				if not rad.item.jewelRadiusData then
+					rad.item.jewelRadiusData = { }
+				end
+				rad.item.jewelRadiusData[rad.nodeId] = rad.data
+			end
+		end
+	end
+	/* */
+
+	return modList
+}
+
+func buildModListForNode(env *Environment, node data.Node) *ModList {
+	var modList = NewModList()
+	for i, stat := range node.Stats {
+		var mods, err = parseMod(stat, i)
+		if strings.Trim(err, " ") != "" {
+			env.DebugErrors = append(env.DebugErrors, "Error parsing Passive Node ("+*node.Name+") mod: "+err+", stat text: "+stat+", with "+strconv.Itoa(len(mods))+" mods found")
+		}
+		for _, mod := range mods {
+			modList.AddMod(mod)
+		}
+	}
```

Comment:
Where in PoB source is this from?

Reply:
That's a good question... I took some liberty with that one. In POB the tree structure already has all the mods available. I should probably do that when the tree is parsed rather than here on every recalculation!